new-new-packages/compiler/src/compiler/ast.rs

```diff
@@ -88,3 +96,144 @@ impl Deref for AstString {
         &self.value
     }
 }
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum AstError {
+    UnexpectedPunctuation,
+    TextWithoutClosingQuote,
+    ParenthesizedWithoutClosingParenthesis,
+    CallOfANonIdentifier,
+    StructWithNonStructField,
+    StructWithoutClosingBrace,
+    ColonMissingAfterStructKey,
+    NonCommaAfterStructValue,
+    ExpectedParameter,
+    LambdaWithoutClosingCurlyBrace,
+}
+
+pub trait CollectErrors {
+    fn collect_errors(self, errors: &mut Vec<CompilerError>);
+}
+impl CollectErrors for Ast {
+    fn collect_errors(self, errors: &mut Vec<CompilerError>) {
+        match self.kind {
+            AstKind::Int(_) => {}
+            AstKind::Text(_) => {}
+            AstKind::Identifier(_) => {}
+            AstKind::Symbol(_) => {}
+            AstKind::Struct(struct_) => {
+                for (key, value) in struct_.fields {
+                    key.collect_errors(errors);
+                    value.collect_errors(errors);
+                }
+            }
+            AstKind::Lambda(lambda) => lambda.body.collect_errors(errors),
+            AstKind::Call(call) => call.arguments.collect_errors(errors),
+            AstKind::Assignment(assignment) => assignment.body.collect_errors(errors),
+            AstKind::Error {
+                child,
+                errors: recovered_errors,
+            } => {
+                errors.append(&mut recovered_errors.clone());
+                child.map(|child| child.collect_errors(errors));
+            }
+        }
+    }
+}
+impl CollectErrors for Vec<Ast> {
+    fn collect_errors(self, errors: &mut Vec<CompilerError>) {
+        for ast in self {
+            ast.collect_errors(errors);
+        }
+    }
+}
+
+impl Display for Ast {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}: ", self.id)?;
+        match &self.kind {
+            AstKind::Int(int) => write!(f, "int {}", int.0),
+            AstKind::Text(text) => write!(f, "text \"{}\"", text.0),
+            AstKind::Identifier(identifier) => write!(f, "identifier {}", identifier.0),
+            AstKind::Symbol(symbol) => write!(f, "symbol {}", symbol.0),
+            AstKind::Struct(struct_) => {
+                write!(
+                    f,
+                    "struct [\n{}\n]",
+                    struct_
+                        .fields
+                        .iter()
+                        .map(|(key, value)| format!("{}: {},", key, value))
+                        .join("\n")
+                        .lines()
+                        .map(|line| format!("  {}", line))
```

Comment:
If the value is multiple lines, all lines should be indented.

For example, we want this:

```
[
  Foo: Bar,
  Baz: [
    Foo: Bar,
  ],
]
```

Instead of this:

```
[
  Foo: Bar,
  Baz: [
  Foo: Bar
],
]
```